node/wireguard: Fix node-to-node encryption inconsistencies in kvstore mode

gandro opened the pull request:
WireGuard-based encryption uses different values when it comes to the encryptKey field in the IPCache:

 - For endpoints, the encryptKey should always be non-zero if pod-to-pod encryption is enabled.
 - For nodes, the encryptKey should only be non-zero if node-to-node encryption is enabled and the node has not opted out of node-to-node encryption via label selector.

Before this PR, we would derive the endpoint encryptKey from the local node store  encryptKey. This however led to inconsistencies, in particular in combination with kvstore-mode. This PR changes that approach slightly and changes the local node store encryptKey to always match the encryptKey of the published CiliumNode or kvstore node resource. 

This fixes three issues:

 - Fixes a bug where opt-out from node-to-node encryption was not working in kvstore mode
 - Fixes a bug where the encryptKey for the node object in the kvstore was flapping if node-to-node encryption was turned off. This did not affect encryption of workloads, as the node encryptKey is only relevant for node-to-node encryption. It did affect health and ingress IPs however (see next item).
 - Fixes a bug where traffic to a node's health and ingress IP were not encrypted in all cases.

giorio94 committed:
node: drop unused SetOptOutNodeEncryption function

This global setter is no longer used since 95029ec3e99d
("daemon: Implement LocalNodeInitializer to fill local node info").

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

gandro committed:
node: Fix inconsistent EncryptKey index handling

WireGuard-based encryption uses different values when it comes to the
EncryptKey field in the IPCache:

 - For endpoints, the EncryptKey should always be non-zero if pod-to-pod
   encryption is enabled.
 - For nodes, the EncryptKey should be non-zero if node-to-node
   encryption is enabled, and if the node has not opted out of
   node-to-node encryption.

Before this commit, we were deriving the EncryptKey of endpoints written
to the kvstore IPCache (see `runIPIdentitySync`) based on the node's
EncryptKey - which is the wrong source of truth for that value.

Luckily, due to a bug, it is currently not possible for Cilium nodes
running in kvstore mode to opt-out of node-to-node encryption, so the
value was always effectively non-zero and the result was accidentally
correct.

However, as we want to fix the node-to-node opt-out mechanism in a
subsequent commit, we ought to fix that first. Therefore, this commit
fixes up all call sites which set an endpoint's EncryptKey to use the
same source of truth.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

giorio94 commented on pkg/node/address.go:
```diff
@@ -574,14 +569,18 @@ func GetIngressIPv6() net.IP {
-		return GetIPsecKeyIdentity()
+		return getLocalNode().EncryptionKey
 	case option.Config.EnableWireguard:
 		if len(GetWireguardPubKey()) > 0 {
```
I guess we could additionally simplify this logic by dropping the `GetWireguardPubKey` check, given that the key is always set if wireguard is enabled (unless an error occurred, but in that case we should have already bailed out).

gandro replied:
Agreed!

giorio94 commented on pkg/wireguard/agent/agent.go:
```diff
@@ -98,7 +98,13 @@ func NewAgent(privKeyPath string, localNodeStore *node.LocalNodeStore) (*Agent,
+			// encryption is disabled, the EncryptionKey field is ignored by
+			// remote nodes (see node/manager.nodeAddressHasEncryptKey).
+			localNode.EncryptionKey = types.StaticEncryptKey
+		}
```
Now that I'm going through this flow once more, I'm pretty confident that this would not work as expected. Indeed, the `localNode.OptOutNodeEncryption` value is initialized by the [initFromK8s](https://github.com/cilium/cilium/blob/355f2ffef29ad30fbdfcbf3ff20b5356e7bc994e/daemon/cmd/local_node_sync.go#L194) function (indirectly) invoked by the [LocalNodeStore start hook](https://github.com/cilium/cilium/blob/355f2ffef29ad30fbdfcbf3ff20b5356e7bc994e/pkg/node/local_node_store.go#L100). But here we are in the context of a provided function, which is executed earlier than start hooks (and thus `optOut` will always be false). 

One solution could be initializing the encryption key as part of `initFromK8s` as well, although that would still leave open how to correctly retrieve the `OptOutNodeEncryption` value for the wireguard module (AFAICT, it seems to be used for status reporting only). Setting the encryption key in a start hook here instead would be prone to race conditions, as observers (e.g., the nodediscovery package) may first see and propagate the node object without the encryption key set.

gandro replied:
Thanks for looking into this. I put the PR back into draft because local testing also revealed that this mechanism was not working.

giorio94 replied:
Yeah, I honestly noticed you put it back to draft only after writing the review :sweat_smile:

giorio94 reviewed (changes requested):
Thanks Sebastian! 

Overall the approach looks good to me. I've one concern about the `OptOutNodeEncryption` logic though, which if I'm not mistaken is currently broken as configured later than when it is retrieved. More details in the inline comment.

And one nit in the commit message: `s/node-to-node is also encryption/node-to-node is also enabled/`

gandro committed:
node: Remove superfluous if condition

This removes an unnecessary if condition in
`GetEndpointEncryptKeyIndex`. If WireGuard is enabled, the WireGuard
agent (see `pkg/wireguard/agent.NewAgent`) always sets a public key
for the local node.

Therefore this if condition is always true and we currently have no
plans for nodes to have WireGuard enabled without them also producing a
public key. Therefore this commit removes that superfluous condition.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
wireguard: Remove unused field member

This commit removes the localNodeStore field from the WireGuard agent,
as is it not accessed outside of the constructor.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
wireguard: Fix node-to-node encryption opt-out in kvstore mode

This commit fixes an issue where the node's EncryptKey field in the
kvstore was not set to zero for nodes which opted out of node-to-node
encryption, thereby effectively encrypting traffic for nodes which
should not have been encrypted.

This was due to the fact that the logic zeroing the local node's
EncryptKey based on `localNode.OptOutNodeEncryption` was only performed
when writing to the K8s CiliumNode CRD (in
`NodeDiscovery.mutateNodeResource`), but not when writing to the kvstore
(in the controller started by `NodeDiscovery.updateLocalNode`).

This commit fixes this issue by zeroing `localNode.EncryptKey` directly
at the source of truth in the `LocalNodeStore`.

In addition, we need to be careful when that value is being written:

 - To avoid flapping values, the EncryptKey value needs to be initialized
   before we publish the local node object to the kvstore or k8s.
   Therefore, initialization of the EncryptKey field needs to happen
   before `LocalNodeSynchronizer.InitLocalNode` returns.

 - In order to determine if the local node has to opt out of
   node-to-node encryption, we need to know the local node's labels.
   Those are only available after `localNodeSynchronizer.initFromK8s` has
   been called - therefore we cannot set the field in a hive constructor
   like we do for some other fields that need to be initialized early.

Therefore, this commit moves the initialization of the WireGuard related
fields into `LocalNodeSynchronizer.InitLocalNode`. This satisfies the
the above constraints for the `EncryptKey` and `OptOutNodeEncryption`
values. The initialization of `WireguardPubKey` and the accompanying
annotation can be performed earlier without loss of correctness, but
to keep all local node initialization in the same place, we now also
write those fields during `InitLocalNode`. This is safe, because
`InitLocalNode` happens before the K8s/kvstore node object is published.

This commit incidentally also fixes a bug where the
`OptOutNodeEncryption` field in the WireGuard agent was read out too
early, i.e. before `InitLocalNode` had a chance to initialize it. That
bug had little consequence, as it only materialized in the `cilium
status` not reporting the opt-out status correctly, and caused node IPs
to be unnecessarily added to the WireGuard peer list. Both issues did
not affect the encrypted traffic.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
node/manager: Fix encryptKey for health/ingress IPs

WireGuard-based encryption uses different values when it comes to the
EncryptKey field in the IPCache:

 - For endpoints, the EncryptKey should always be non-zero if pod-to-pod
   encryption is enabled.
 - For nodes, the EncryptKey should be non-zero if node-to-node
   encryption is enabled, and if the node has not opted out of
   node-to-node encryption.

When creating IPCache entries for regular endpoints, the EncryptKey
value is taken from either the CEP/CES Kubernetes custom resource, or
from the IP entry in the kvstore.

However, the IPCache entries for the health and ingress endpoints are
dervied from the node resource. Before this commit, those entries
therefore also used the node's EncryptKey. When a node opted out of
node-to-node encryption however, that meant that we did not encrypt
traffic for health and ingress to those nodes.

Thus, this commit works around that issue by ignoring the node's
EncryptKey for the health and ingress endpoints and instead always
encrypt that traffic (which is what we do for regular endpoints with
WireGuard encryption already).

This commit can be considered a workaround. Ideally, we would have
separate fields in the node resource for the node and its special
endpoints. That however is a larger schema change, thus this commit
focuses on backportable changes.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

giorio94 commented on daemon/cmd/local_node_sync.go:
```diff
@@ -54,6 +55,11 @@ func (ini *localNodeSynchronizer) InitLocalNode(ctx context.Context, n *node.Loc
 		return err
 	}
+
+	if ini.WireGuard != nil {
```
This check should not be necessary, as WireGuard is not an optional dependency.

gandro replied:
So if WireGuard is disabled, we return a nil-valued pointer here:
https://github.com/cilium/cilium/blob/835e9613d43a4ebbeb6a1dfafcfd4238cbe29bc8/pkg/datapath/cells.go#L155

I guess there is a difference between a `nil` value being provided (which can be done for required values) and no value being provided (which only works for optional values).

giorio94 replied:
Right, I missed it could legitimately be nil.

giorio94 reviewed (approved):
Thanks! Changes look good to me pending the fix for the tests.
And kudos for the amazing commit messages!

aditighag reviewed (approved):
Looks good for my codeowners cilium/endpoint.

brb reviewed (approved):
:+1: Thanks for nice commit msgs.
